File: app/src/main/java/org/nevack/unitconverter/converter/ConverterFragment.java
package org.nevack.unitconverter.converter;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import com.google.android.material.snackbar.Snackbar;

import org.nevack.unitconverter.R;
import org.nevack.unitconverter.converter.ConverterContract.ConvertData;
import org.nevack.unitconverter.history.HistoryActivity;
import org.nevack.unitconverter.model.Unit;

import java.util.List;

public class ConverterFragment extends Fragment implements ConverterContract.View {

    private KeypadView keypadView;
    private ConverterDisplayView displayView;
    private View background;

    private ConverterContract.Presenter presenter;

    public ConverterFragment() {
        super(R.layout.fragment_converter);
    }

    public static ConverterFragment newInstance() {
        return new ConverterFragment();
    }

    @Override
    public void setPresenter(@NonNull ConverterContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.start();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        final Toolbar toolbar = view.findViewById(R.id.toolbar);
        final AppCompatActivity activity = (AppCompatActivity) requireActivity();
        activity.setSupportActionBar(toolbar);

        final ActionBar actionBar = activity.getSupportActionBar();
        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(R.drawable.ic_menu);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        keypadView = view.findViewById(R.id.keypad);
        displayView = view.findViewById(R.id.display);

        displayView.setTextWatcher(new TextChangedWatcher(s -> presenter.convert(displayView.getConvertData())));

        displayView.setSpinnersCallback(() -> presenter.convert(displayView.getConvertData()));

        background = view.findViewById(R.id.background);

        keypadView.setOnCopyListeners(
                new KeypadView.ActionListener() {
                    @Override
                    public void longClick() {
                        copyWithNotification(true);
                    }

                    @Override
                    public void singleClick() {
                        copyWithNotification(false);
                    }
                }
        );

        keypadView.setBackspaceListeners(
                new KeypadView.ActionListener() {
                    @Override
                    public void longClick() {
                        displayView.erase();
                    }

                    @Override
                    public void singleClick() {
                        displayView.removeLastDigit();
                    }
                }
        );

        keypadView.setOnPasteListener(v -> presenter.pasteFromClipboard());

        keypadView.setNumericListener(number -> displayView.appendText(String.valueOf(number)));
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_converter, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.save:
                presenter.saveResultToHistory();
                break;
            case R.id.history:
                startActivity(new Intent(requireContext(), HistoryActivity.class));
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void setTitle(int title) {
        requireActivity().setTitle(title);
    }

    @Override
    public void setUnits(List<Unit> units) {
        displayView.setUnits(units);
    }

    @Override
    public ConvertData getConvertData() {
        return displayView.getConvertData();
    }

    @Override
    public void setConvertData(ConvertData data) {
        displayView.setConvertData(data);
    }

    @Override
    public void appendText(String text) {
        displayView.appendText(text);
    }

    @Override
    public void showResult(String result) {
        displayView.showResult(result);
    }

    @Override
    public void clear() {
        displayView.clear();
    }

    @Override
    public void showError() {
        displayView.showError();
    }

    @Override
    public void setBackgroundColor(int color) {
        background.setBackgroundColor(ContextCompat.getColor(requireContext(), color));
    }

    private void copyWithNotification(boolean shouldIncludeUnit) {
        presenter.copyResultToClipboard(displayView.getCopyResult(shouldIncludeUnit));
        Snackbar.make(background, R.string.copy_result_toast, Snackbar.LENGTH_SHORT).show();
    }

    private static class TextChangedWatcher implements TextWatcher {

        private interface Callback {
            void onTextChanged(String netText);
        }

        private Callback callback;

        TextChangedWatcher(Callback callback) {
            this.callback = callback;
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            // No-op
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            // No-op
        }

        @Override
        public void afterTextChanged(Editable s) {
            callback.onTextChanged(s.toString());
        }
    }
}

File: app/src/main/java/org/nevack/unitconverter/converter/ConverterFragment.kt
package org.nevack.unitconverter.converter

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.google.android.material.snackbar.Snackbar
import dev.chrisbanes.insetter.applyInsetter
import org.nevack.unitconverter.R
import org.nevack.unitconverter.converter.ConverterContract.ConvertData
import org.nevack.unitconverter.databinding.FragmentConverterBinding
import org.nevack.unitconverter.history.HistoryActivity

class ConverterFragment : Fragment(R.layout.fragment_converter), ConverterContract.View {
    private lateinit var binding: FragmentConverterBinding

    private val viewModel: ConverterViewModel by activityViewModels()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        binding = FragmentConverterBinding.bind(view)

        binding.toolbar.inflateMenu(R.menu.menu_converter)
        binding.toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.save -> {
                    viewModel.saveResultToHistory()
                    true
                }
                R.id.history -> {
                    startActivity(Intent(requireContext(), HistoryActivity::class.java))
                    true
                }
                else -> false
            }
        }
        binding.toolbar.setNavigationIcon(R.drawable.ic_menu)
        binding.toolbar.setNavigationOnClickListener {
            viewModel.setDrawerOpened(true)
        }

        binding.display.setTextWatcher {
            viewModel.convert(binding.display.convertData)
        }
        binding.display.setSpinnersCallback {
            viewModel.convert(binding.display.convertData)
        }
        binding.keypad.setOnCopyListeners(
            object : KeypadView.ActionListener {
                override fun longClick() {
                    copyWithNotification(true)
                }

                override fun singleClick() {
                    copyWithNotification(false)
                }
            }
        )
        binding.converterDisplayContainer?.applyInsetter {
            type(WindowInsetsCompat.Type.statusBars()) { padding() }
        }
        binding.keypad.applyInsetter {
            type(WindowInsetsCompat.Type.systemBars()) { padding() }
        }
        binding.keypad.setBackspaceListeners(
            object : KeypadView.ActionListener {
                override fun longClick() {
                    binding.display.erase()
                }

                override fun singleClick() {
                    binding.display.removeLastDigit()
                }
            }
        )
        binding.keypad.setOnPasteListener {
//            presenter!!.pasteFromClipboard()
        }
        binding.keypad.setNumericListener { number -> binding.display.appendText(number.toString()) }

        viewModel.backgroundColor.observe(viewLifecycleOwner) {
            binding.background.setBackgroundColor(ContextCompat.getColor(requireContext(), it))
        }

        viewModel.title.observe(viewLifecycleOwner) {
            binding.toolbar.setTitle(it)
        }

        viewModel.converter.observe(viewLifecycleOwner) {
            binding.display.setUnits(it.units)
        }

        viewModel.result.observe(viewLifecycleOwner) {
            binding.display.showResult(it)
        }
    }

    override fun getConvertData(): ConvertData = binding.display.convertData

    override fun setConvertData(data: ConvertData) {
        binding.display.convertData = data
    }

    override fun appendText(digit: String) = binding.display.appendText(digit)

    override fun clear() = binding.display.clear()

    override fun showError() = binding.display.showError()

    private fun copyWithNotification(shouldIncludeUnit: Boolean) {
        viewModel.copyResultToClipboard(binding.display.getCopyResult(shouldIncludeUnit))
        Snackbar.make(binding.background, R.string.copy_result_toast, Snackbar.LENGTH_SHORT).show()
    }
}
